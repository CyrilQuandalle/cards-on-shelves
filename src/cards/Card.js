'use strict'

import React from 'react';
import classNames from 'classnames';

export class Card extends React.Component {
    constructor(props) {
        super(props);
        this.card = props.card;
        this.cardItems = props.card.content || [];
    }

    render() {
        const card = this.card;
        const shapeClasses = classNames('card-shape',
            {'landscape': 'landscape' == card.shape}
        );
        const cardClasses = classNames('card');
        const cardItems = this.cardItems.map((item, index) =>
            <CardItem key={index} item={item} />
        );
        return (
            <div className={shapeClasses}>
                <div className={cardClasses}>
                    {cardItems}
                </div>
            </div>
        );
    }
}

function CardItem(props){
    let item = props.item;
    if('title' == item.type) {
        return <CardTitle item={item} />;
    }
    if('field' == item.type) {
        return <CardField item={item} />;
    } 
    return <CardText item={item} />;
}

function CardTitle(props) {
    const item = props.item;
    return (
        <div className="card-item emphasized short">
            <span className="card-header-name">{item.content}</span>
        </div>
    );
}

function CardField(props) {
    let item = props.item;
    return (
        <div className="card-item short no-padding">
            <div className="card-field">
                <div className="card-field-header">{item.header}</div>
                <div className="card-field-value">{item.value}</div>
            </div>
        </div>
    );
}

function CardText(props) {
    const item = props.item;
    return (<div className="card-item card-long-text">{item.content}</div>);
}
