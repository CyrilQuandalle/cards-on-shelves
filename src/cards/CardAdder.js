'use strict'

import React from 'react';

export class CardAdder extends React.Component {
    constructor(props) {
        super(props);
        this.onAdd = props.onAdd; 
    }

    render() {
        return (
            <button type="button" onClick={ this.onAdd }>+</button>
        );
    }
}
