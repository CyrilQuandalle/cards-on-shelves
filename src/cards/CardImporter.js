'use strict'

import React from 'react';

export class CardImporter extends React.Component {
    constructor(props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
//        this.onImport = props.onImport;
        this.fileImporter = this.createFileImporter(props.onImport);
    }
    
    createFileImporter(onImport) {
        let reader = new FileReader();
        reader.onload = function (evt) {
            try {
                let cards = JSON.parse(evt.target.result);
                onImport(cards);
            } catch (e) {
                throw e;
            }
        }
        reader.onerror = function (evt) {
            console.error('Error reading file');
        }
        return reader;
    }

    render() { 
        return (
            <input type="file" onChange={(e) => this.handleChange(e.target.files) } />
        );
    }
    
    handleChange(selectorFiles: FileList) {
        const importer = this;
        Array.from(selectorFiles).forEach(file => this.importFile(file));
    }

    importFile(file) {
        //const importer = this;

        //let reader = new FileReader();
        this.fileImporter.readAsText(file, "UTF-8");
/*        reader.onload = function (evt) {
            try {
                let cards = JSON.parse(evt.target.result);
                importer.onImport(cards);
            } catch (e) {
                throw e;
            }
        }
        reader.onerror = function (evt) {
            console.error('Error reading file');
        }*/
    }
}
