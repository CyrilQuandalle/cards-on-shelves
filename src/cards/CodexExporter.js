'use strict'

import React from 'react';
import FileSaver from 'file-saver';

export class CodexExporter extends React.Component {
    constructor(props) {
        super(props);
        this.codex = props.codex;
        this.exportCards = this.exportCards.bind(this);
    }

    render() {
        return (
            <button type="button" onClick={ this.exportCards }>Sauver</button>
        );
    }
    
    exportCards() {
        const codex = this.codex;
        let blob = new Blob([JSON.stringify(codex.cards)], {type: "text/plain;charset=utf-8"});
        FileSaver.saveAs(blob, codex.name + ".json");
    }
}
