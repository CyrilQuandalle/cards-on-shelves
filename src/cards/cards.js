'use strict'

export function card(card) {
    var shape = document.createElement('div');
    shape.classList.add('card-shape');

    var cardElement = document.createElement('div');
    cardElement.classList.add('card');

    if(card && card.content){
        if('landscape' == card.type) {
            shape.classList.add('landscape');
        }


        for (let item of card.content) {
            let itemElement = createItemElement(item);
            cardElement.appendChild(itemElement);
        }
    }

    shape.appendChild(cardElement); 
    return shape;
}

function createItemElement(item){
    if('title' == item.type) {
        return createTitle(item);
    }
    if('field' == item.type) {
        return createField(item);
    } 
    return createLongText(item);
}

function createTitle(title) {
    let item = document.createElement('div');
    item.classList.add('card-item');
    item.classList.add('emphasized');
    item.classList.add('short');
    
    let titleName = document.createElement('span');
    titleName.classList.add('card-header-name');
    titleName.innerHTML = title.content;

    item.appendChild(titleName);
    return item;
}

function createLongText(text) {
    let item = document.createElement('div');
    item.classList.add('card-item');
    item.classList.add('card-long-text');
    item.innerHTML = text.content;
    return item;
}

function createField(field) {
    let item = document.createElement('div');
    item.classList.add('card-item');
    item.classList.add('short');
    item.classList.add('no-padding');
    
    let header = document.createElement('div');
    header.classList.add('card-field-header');
    header.innerHTML = field.header;
    
    let value = document.createElement('div');
    value.classList.add('card-field-value');
    value.innerHTML = field.value;

    let fieldElement = document.createElement('div');
    fieldElement.classList.add('card-field'); 
    fieldElement.appendChild(header);
    fieldElement.appendChild(value);

    item.appendChild(fieldElement);
    return item;
}
