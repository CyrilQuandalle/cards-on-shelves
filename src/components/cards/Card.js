'use strict';

import React, {Component} from 'react';
import classNames from 'classnames';

import {CardItemContainer} from './items/CardItemContainer.js';

import styles from './card.css';

export class Card extends Component {
    constructor(props) {
        super(props);
        this.card = props.card;
        this.cardItems = props.card.content || [];
    }

    render() {
        const card = this.card;
        const shapeClasses = classNames.bind(styles)('card-shape',
            {'landscape': 'landscape' === card.shape}
        );
        return (
            <div className={shapeClasses}>
                <CardItemContainer cardItems={this.cardItems}/>
            </div>
        );
    }
}
