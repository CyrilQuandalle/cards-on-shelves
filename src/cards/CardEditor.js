'use strict'

import React from 'react';
import classNames from 'classnames';

import { Card } from './Card.js';

export class CardEditor extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            card: props.card
        };
        this.handleCardChange = this.handleCardChange.bind(this);
        this.updateItem = this.updateItem.bind(this);
    }

    render() {
        const card = this.state.card;
        const cardItems = card.content || [];
        const cardItemInputs = cardItems.map((item, index) =>
            <CardItemInput key={index} item={item} onUpdate={ this.updateItem }/>
        );

        return (
            <div className="card-editor-container">
                <div className="sheet-name">{card.name}</div>
                <Card card={card} />
                <form className="card-editor">
                    <input name="name" type="text" value={card.name} onChange={ this.handleCardChange } />
                    {false &&
                    <div className="card-editor-toolbar">
                        <button>Titre</button>
                        <button>Zone de texte</button>
                        <button>Cl√©/Valeur</button>
                    </div>
                    }
                    <fieldset>
                        {cardItemInputs}
                    </fieldset>
                </form>
            </div>
        );
    }

    handleCardChange(event) {
        const target = event.target;
        const value = target.value;
        const name = target.name;

        let card = this.state.card;
        card[name] = value;

        this.updateCard(card);
    }

    updateCard(card) {
        this.setState({card: card});
    }

    updateItem(item) {
        this.refresh();
    }

    refresh() {
        let card = this.state.card;
        this.updateCard(card);
    }
}

function CardItemInput(props){
    let item = props.item;
    if('title' == item.type) {
        return <CardTitleInput item={item} onUpdate={props.onUpdate}/>;
    }
    if('field' == item.type) {
        return <CardFieldInput item={item} onUpdate={props.onUpdate} />;
    } 
    return <CardTextInput item={item} onUpdate={props.onUpdate} />;
}

class CardTitleInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            item: props.item
        };
        this.handleTitleChange = this.handleTitleChange.bind(this);
        this.onUpdate = props.onUpdate;
    }

    render() {
        const item = this.state.item;
        return (
            <input type="text" value={item.content} onChange={ this.handleTitleChange } />
        );
    }
    
    handleTitleChange(event) {
        let item = this.state.item;
        item.content = event.target.value;

        this.updateItem(item)
    }

    updateItem(item){
        this.setState({
            item: item
        });
        this.onUpdate(item);
    }
}

class CardFieldInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            item: props.item
        };
        this.handleHeaderChange = this.handleHeaderChange.bind(this);
        this.handleValueChange = this.handleValueChange.bind(this);
        this.onUpdate = props.onUpdate;
    }

    render() {
        let item = this.state.item;
        return (
            <div className="card-input-field">
                <input type="text" value={item.header} onChange={ this.handleHeaderChange } />
                <input type="text" value={item.content} onChange={ this.handleValueChange } />
            </div>
        );
    }

    handleHeaderChange(event) {
        let item = this.state.item;
        item.header = event.target.value;

        this.updateItem(item)
    }

    handleValueChange(event) {
        let item = this.state.item;
        item.content = event.target.value;

        this.updateItem(item)
    }

    updateItem(item){
        this.setState({
            item: item
        });
        this.onUpdate(item);
    }

}

class CardTextInput extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            item: props.item
        };
        this.handleTextChange = this.handleTextChange.bind(this);
        this.onUpdate = props.onUpdate;
    }

    render() {
        const item = this.state.item;
        return (
            <textarea value={item.content} onChange={ this.handleTextChange } />
        );
    }
    
    handleTextChange(event) {
        let item = this.state.item;
        item.content = event.target.value;

        this.updateItem(item)
    }

    updateItem(item){
        this.setState({
                        item: item
                    });
        this.onUpdate(item);
    }
}
