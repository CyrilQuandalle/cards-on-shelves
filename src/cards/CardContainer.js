'use strict';

import React, {Component} from 'react';

import {Card} from './Card.js';
import {CardEditor} from './CardEditor.js';
import {CardImporter} from './CardImporter.js';

export class CardContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            cards: props.cards
        };
        this.createCard = this.createCard.bind(this);
        this.removeCard = this.removeCard.bind(this);
        this.addCard = this.addCard.bind(this);
        this.addCards = this.addCards.bind(this);
        this.onUpdateContainer = props.onUpdateContainer;
    }

    updateContainer(cards) {
        this.setState({cards: cards});
        this.onUpdateContainer(cards);
    }

    createCard() {
        const aNewCard = {
            "name": "",
            "content": []
        };
        this.addCard(aNewCard);
    }

    addCard(card) {
        this.addCards([card]);
    }

    addCards(cards) {
        this.updateContainer(this.state.cards.concat(...cards));
    }

    removeCard(card) {
        let cards = this.state.cards;
        let cardIndex = cards.indexOf(card);
        cards.splice(cardIndex, 1);
        this.updateContainer(cards);
    }

    render() {
        let cards = this.state.cards;
        const items = cards.map((card, index) =>
            <CardContainerItem key={index} card={card} addCard={this.addCard} removeCard={this.removeCard}/>
        );
        return (
            <div className="card-container">
                <div className="card-container-toolbar">
                    <button type="button" onClick={this.createCard}>+</button>
                    <CardImporter onImport={this.addCards}/>
                </div>
                <div className="card-container-items">
                    {items}
                </div>
            </div>);
    }
}

class CardContainerItem extends Component {
    constructor(props) {
        super(props);
        this.state = {
            card: props.card
        };
        this.addCard = props.addCard;
        this.removeCard = props.removeCard;
        this.onUpdateCard = this.onUpdateCard.bind(this);
    }

    onUpdateCard(card) {
        this.setState({card: card});
    }

    render() {
        let card = this.state.card;
        return (
            <div className="card-container-item">
                <div className="card-container-item-name">{card.name}</div>
                <div className="card-container-item-toolbar">
                    <button type="button" onClick={() => this.addCard(card)}>Dupliquer</button>
                    <button type="button" onClick={() => this.removeCard(card)}>Supprimer</button>
                </div>
                <Card card={card}/>
                <CardEditor card={card} onUpdate={this.onUpdateCard}/>
            </div>
        )
    }
}
