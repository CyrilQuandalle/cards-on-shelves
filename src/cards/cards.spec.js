'use strict'

import * as cards from './cards.js';
import fs from 'fs';
import beautify from 'js-beautify';

describe("A single card", () => {
    it('could be empty', () => {
        var card = {};
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_card_empty.html'));
    });
    
    it('should render title', () => {
        var card = {
            content: [
                {
                    type: 'title',
                    content: 'Titre de la carte.'
                }
            ]
        };
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_card_with_header.html'));
    });

    it('should render with long text', () => {
        var card = {
            content: [
                {
                    type: 'long-text',
                    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam dui magna, tristique ac volutpat sed, sollicitudin molestie magna. Cras sed purus at metus viverra venenatis a et dolor. Pellentesque egestas nisi in mollis aliquam. Sed scelerisque fermentum tincidunt. Nulla interdum venenatis sapien, sit amet condimentum lectus. Phasellus malesuada semper quam, ut pulvinar dolor iaculis vitae. Praesent eu lorem quis nisl lobortis venenatis non id massa. Curabitur non ex sodales, mollis tellus in, dapibus purus. Maecenas eget odio commodo, posuere sapien eget, congue arcu. Donec id facilisis lectus. Proin vitae ipsum libero. Quisque vitae tellus dui. Nulla tincidunt est ac quam faucibus, a lobortis elit venenatis. Sed quis neque sit amet leo finibus cursus. Aenean gravida consectetur erat, sit amet lobortis magna volutpat iaculis.'
                }
            ]
        };
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_card_with_long_text.html'));
    });

    it('should render with title and long text', () => {
        var card = {
            content: [
                 {
                    type: 'title',
                    content: 'Titre de la carte.'
                },
                {
                    type: 'long-text',
                    content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam dui magna, tristique ac volutpat sed, sollicitudin molestie magna. Cras sed purus at metus viverra venenatis a et dolor. Pellentesque egestas nisi in mollis aliquam. Sed scelerisque fermentum tincidunt. Nulla interdum venenatis sapien, sit amet condimentum lectus. Phasellus malesuada semper quam, ut pulvinar dolor iaculis vitae. Praesent eu lorem quis nisl lobortis venenatis non id massa. Curabitur non ex sodales, mollis tellus in, dapibus purus. Maecenas eget odio commodo, posuere sapien eget, congue arcu. Donec id facilisis lectus. Proin vitae ipsum libero. Quisque vitae tellus dui. Nulla tincidunt est ac quam faucibus, a lobortis elit venenatis. Sed quis neque sit amet leo finibus cursus. Aenean gravida consectetur erat, sit amet lobortis magna volutpat iaculis.'
                }
            ]
        };
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_card.html'));
    });

    it('should be rendered with field', () => {
        var card = {
            content: [
                {
                    type: 'field',
                    header: 'Type de carte',
                    value: 'MJ'
                 }
            ]
        };
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_card_with_field.html'));
    });

});

describe("A landscape card", () => {
    it('should be rendered with field', () => {
        var card = {
            type: 'landscape',
            content: [
                {
                    type: 'field',
                    header: 'Type de carte',
                    value: 'MJ'
                 }
            ]
        };
        var rendered = cards.card(card);

        expect(beautify.html(rendered.outerHTML))
            .toBe(expectedFile('single_gray_landscape_card_with_field.html'));
    });
});

function expectedFile(filename){
        return fs.readFileSync('./src/cards/expects/'+ filename, 'utf-8').trim();
}
