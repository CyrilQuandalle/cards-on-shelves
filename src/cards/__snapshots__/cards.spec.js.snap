// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A single Component card could be empty 1`] = `
<div
  className="card-shape"
>
  <div
    className="card"
  />
</div>
`;

exports[`A single Component card could be landscaped 1`] = `
<div
  className="card-shape landscape"
>
  <div
    className="card"
  >
    <div
      className="card-item short no-padding"
    >
      <div
        className="card-field"
      >
        <div
          className="card-field-header"
        >
          Type de carte
        </div>
        <div
          className="card-field-value"
        >
          MJ
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`A single Component card should be rendered with field 1`] = `
<div
  className="card-shape"
>
  <div
    className="card"
  >
    <div
      className="card-item short no-padding"
    >
      <div
        className="card-field"
      >
        <div
          className="card-field-header"
        >
          Type de carte
        </div>
        <div
          className="card-field-value"
        >
          MJ
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`A single Component card should render title 1`] = `
<div
  className="card-shape"
>
  <div
    className="card"
  >
    <div
      className="card-item emphasized short"
    >
      <span
        className="card-header-name"
      >
        Titre de la carte.
      </span>
    </div>
  </div>
</div>
`;

exports[`A single Component card should render with long text 1`] = `
<div
  className="card-shape"
>
  <div
    className="card"
  >
    <div
      className="card-item card-long-text"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam dui magna, tristique ac volutpat sed, sollicitudin molestie magna. Cras sed purus at metus viverra venenatis a et dolor. Pellentesque egestas nisi in mollis aliquam. Sed scelerisque fermentum tincidunt. Nulla interdum venenatis sapien, sit amet condimentum lectus. Phasellus malesuada semper quam, ut pulvinar dolor iaculis vitae. Praesent eu lorem quis nisl lobortis venenatis non id massa. Curabitur non ex sodales, mollis tellus in, dapibus purus. Maecenas eget odio commodo, posuere sapien eget, congue arcu. Donec id facilisis lectus. Proin vitae ipsum libero. Quisque vitae tellus dui. Nulla tincidunt est ac quam faucibus, a lobortis elit venenatis. Sed quis neque sit amet leo finibus cursus. Aenean gravida consectetur erat, sit amet lobortis magna volutpat iaculis.
    </div>
  </div>
</div>
`;

exports[`A single Component card should render with title and long text 1`] = `
<div
  className="card-shape"
>
  <div
    className="card"
  >
    <div
      className="card-item emphasized short"
    >
      <span
        className="card-header-name"
      >
        Titre de la carte.
      </span>
    </div>
    <div
      className="card-item card-long-text"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam dui magna, tristique ac volutpat sed, sollicitudin molestie magna. Cras sed purus at metus viverra venenatis a et dolor. Pellentesque egestas nisi in mollis aliquam. Sed scelerisque fermentum tincidunt. Nulla interdum venenatis sapien, sit amet condimentum lectus. Phasellus malesuada semper quam, ut pulvinar dolor iaculis vitae. Praesent eu lorem quis nisl lobortis venenatis non id massa. Curabitur non ex sodales, mollis tellus in, dapibus purus. Maecenas eget odio commodo, posuere sapien eget, congue arcu. Donec id facilisis lectus. Proin vitae ipsum libero. Quisque vitae tellus dui. Nulla tincidunt est ac quam faucibus, a lobortis elit venenatis. Sed quis neque sit amet leo finibus cursus. Aenean gravida consectetur erat, sit amet lobortis magna volutpat iaculis.
    </div>
  </div>
</div>
`;
